Improve code coverage by modifing/adding assertions in GOTO

ChenfengWei0 opened the pull request:
This PR includes:
- Bug fixes for multi-property check: In `multi-property` and `parallel-solving` modes, esbmc "forgets" to output the **"VERIFICATION SUCCESSFUL"** message in the end when no bugs are found.

```c
Symex completed in: 0.001s (47 assignments)
Slicing time: 0.000s (removed 47 assignments)
Generated 35 VCC(s), 0 remaining after simplification (0 assignments)
BMC program time: 0.001s

VERIFICATION SUCCESSFUL  /* This line did not output */
```
- Add three options to modify/add assertions in the GOTO program
  - add an assert(0)  before each `RETRUN`/`END_FUNCTION` statement
  - convert every assertion to an assert(0)
```
  --add-false-assert               add a false assertion before each
                                   RETRUN/END_FUNCTION statement
  --make-assert-false              convert every assertion to false
  --goto-coverage                  this activates --add-false-assert and
                                   --make-assert-false
```

- **Result**
<img width="1024" alt="1" src="https://github.com/esbmc/esbmc/assets/105888669/7d4673dc-6bda-4ea1-bd47-e6903d96de01">

The idea is to combine `--goto-coverage` and `--multi-property` to increase the code coverage. Theoretically, any program using `--goto-coverage` will get a **"VERIFICATION FAILED"** result.

ChenfengWei1 committed:
add goto_coverage

ChenfengWei1 committed:
bug fix

ChenfengWei1 committed:
add options for debug

ChenfengWei1 committed:
add another 2 options for debug

ChenfengWei1 committed:
add description for goto-coverage

ChenfengWei1 committed:
add regression teset cases

ChenfengWei1 committed:
bug fix

ChenfengWei1 committed:
fix regression test

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
fix typo error

ChenfengWei1 committed:
fix codacy

lucasccordeiro reviewed (approved):
Thanks for submitting this PR, @ChenfengWei0.

@rafaelsamenezes: could you please review this PR and check whether it implements the suggestions from Intel?

rafaelsamenezes commented on src/goto-programs/goto_coverage.cpp:
```diff
@@ -0,0 +1,37 @@
+      goto_programt &goto_program = f_it->second.body;
+      Forall_goto_program_instructions(it, goto_program)
+      {
+        if(it->is_return() || it->is_end_function())
```
Maybe add at the beginning and every branching instead of the return

rafaelsamenezes reviewed (changes requested):
Nice ideia overall, I'm not sure if you would miss some branches from the original C though.

I would suggest:

1. Adding a coverage report. It can be as simple as the number of failed divided by all Assertions.

2. I am not sure whether this is affecting the operational models. If so, it shouldn't.

ChenfengWei1 committed:
Modified the strategy for inserting assertions

ChenfengWei1 committed:
change the progress order

ChenfengWei1 committed:
add instrumentation

ChenfengWei1 committed:
add coverage info and beautify output

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
lint code base

ChenfengWei1 committed:
bug fix: allow 0 assertion found

ChenfengWei1 committed:
fix typo error

ChenfengWei0 commented:
Update: 
- Changed the strategy:
  - insert a false assertion at the beginning of each function/branch and the end of each function
  - convert every assertion to false

<img width="932" alt="3" src="https://github.com/esbmc/esbmc/assets/105888669/a0875511-46df-45e8-b6a2-f8ae8f90cb0b">

- Added a coverage report:
  - To get a meaningful coverage number, I think it should always work with the `--no-library` option
<img width="779" alt="1" src="https://github.com/esbmc/esbmc/assets/105888669/a9c41618-52cb-437a-b39b-b70251fa834d">

The untracked assertions could be the uncovered branches or functions.

- disable the `report_multi_property_trace` function temporarily because
  - Some output message is duplicated
  - In `--parallel-solving` mode, the output could be unordered and messy
 
<img width="869" alt="2" src="https://github.com/esbmc/esbmc/assets/105888669/d5d77904-a2ce-45ac-9d61-c1803bcf81ab">

rafaelsamenezes commented:
I would suggest checking some classic Coverage Criteria topics. @lucasccordeiro lecture https://ssvlab.github.io/lucasccordeiro/courses/2020/01/software-security/slides/lecture05.pdf specifies some coverage strategies. I am mentioning this because your current approach is a bit non-standard right now (it seems a mix of Decision and Branch coverage).

ChenfengWei0 commented:
> I would suggest checking some classic Coverage Criteria topics. @lucasccordeiro lecture https://ssvlab.github.io/lucasccordeiro/courses/2020/01/software-security/slides/lecture05.pdf specifies some coverage strategies. I am mentioning this because your current approach is a bit non-standard right now (it seems a mix of Decision and Branch coverage).

Thanks for the information🫡

lucasccordeiro commented:
Chenfeng knows these slides well :-) Shouldn't be a problem to implement
these coverage criteria ;-)

Em qui., 1 de jun. de 2023 às 10:55, Chenfeng Wei ***@***.***>
escreveu:

> I would suggest checking some classic Coverage Criteria topics.
> @lucasccordeiro <https://github.com/lucasccordeiro> lecture
> https://ssvlab.github.io/lucasccordeiro/courses/2020/01/software-security/slides/lecture05.pdf
> specifies some coverage strategies. I am mentioning this because your
> current approach is a bit non-standard right now (it seems a mix of
> Decision and Branch coverage).
>
> Thanks for the information🫡
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/esbmc/esbmc/pull/1048#issuecomment-1571728470>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA4F4HLXXCAPCEEDVFFX5UDXJBRKHANCNFSM6AAAAAAYUKIC4Y>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Lucas C. Cordeiro​

https://ssvlab.github.io/lucasccordeiro/

ChenfengWei0 commented:
I'm a little confused now🤔

The original idea is : 
1) Converting all assertions into false statements (i.e., assert(x) => assert(0)). 
  1.1 Add such assertions for branches, loop conditions etc.
2) Applying multi-property verification so each VCC is verified independently. 
3) Check which assertions fail. If an assertion did not fail,  then it is not reachable.

So now we have an assertion coverage for `3.`, which is `reached_asserts`/`total_asserts`. Do we still need the other coverage metrics?

lucasccordeiro commented:
@rafaelsamenezes: do you have any comments here?

rafaelsamenezes commented:
> I'm a little confused now🤔
> 
> The original idea is :
> 
>     1. Converting all assertions into false statements (i.e., assert(x) => assert(0)).
>        1.1 Add such assertions for branches, loop conditions etc.
> 
>     2. Applying multi-property verification so each VCC is verified independently.
> 
>     3. Check which assertions fail. If an assertion did not fail,  then it is not reachable.
> 
> 
> So now we have an assertion coverage for `3.`, which is `reached_asserts`/`total_asserts`. Do we still need the other coverage metrics?

That's the thing, if we are going only for reachability then 1.1 does not make sense. If we are targeting a specific type of coverage (like cover-branches) then 1.1 should reflect that and we should only count branching assertions (and we shouldn't instrument OMs).

 My sugestion is:

1. Leave 1.1 optional and ignore it for now. Keep 1 as it is.
2. In 3 just use the `reached_assert/total_asserts`.

ChenfengWei1 committed:
fix coverage info

ChenfengWei1 committed:
bug fix

ChenfengWei0 commented:
> > I'm a little confused now🤔
> > The original idea is :
> > ```
> > 1. Converting all assertions into false statements (i.e., assert(x) => assert(0)).
> >    1.1 Add such assertions for branches, loop conditions etc.
> > 
> > 2. Applying multi-property verification so each VCC is verified independently.
> > 
> > 3. Check which assertions fail. If an assertion did not fail,  then it is not reachable.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So now we have an assertion coverage for `3.`, which is `reached_asserts`/`total_asserts`. Do we still need the other coverage metrics?
> 
> That's the thing, if we are going only for reachability then 1.1 does not make sense. If we are targeting a specific type of coverage (like cover-branches) then 1.1 should reflect that and we should only count branching assertions (and we shouldn't instrument OMs).
> 
> My sugestion is:
> 
>     1. Leave 1.1 optional and ignore it for now. Keep 1 as it is.
> 
>     2. In 3 just use the `reached_assert/total_asserts`.

Update: 
- we have `make-assert-false` option for 1
- we have `add-false-assert` option for 1.1 
- Output assertion coverage ( `reached_asserts`/`total_asserts`) when `make-assert-false` is the only enabled option

![image](https://github.com/esbmc/esbmc/assets/105888669/b4a4f184-036a-418d-b42d-59861117918a)

lucasccordeiro commented:
Many thanks, @ChenfengWei0.

I will merge this PR once the CI is green.

ChenfengWei1 committed:
try to resolve conflict

ChenfengWei0 committed:
Merge branch 'master' into test_coverage

lucasccordeiro commented:
Thanks for submitting this PR, @ChenfengWei0.
